src/lightning/pytorch/trainer/configuration_validator.py

```diff
@@ -152,16 +153,20 @@ def __verify_manual_optimization_support(trainer: "pl.Trainer", model: "pl.Light
         )
 
 
-def __check_training_step_requires_dataloader_iter(model: "pl.LightningModule") -> None:
-    """Check if the current `training_step` is requesting `dataloader_iter`."""
-    if is_param_in_hook_signature(model.training_step, "dataloader_iter", explicit=True):
-        for hook in ("on_train_batch_start", "on_train_batch_end"):
-            if is_overridden(hook, model):
-                rank_zero_warn(
-                    f"The `batch_idx` argument in `{type(model).__name__}.{hook}` hook may"
-                    " not match with the actual batch index when using a `dataloader_iter`"
-                    " argument in your `training_step`."
-                )
+def __warn_dataloader_iter_limitations(model: "pl.LightningModule") -> None:
+    """Check if `dataloader_iter is enabled`."""
+    if any(
+        is_param_in_hook_signature(step_fn, "dataloader_iter", explicit=True)
+        for step_fn in (model.training_step, model.validation_step, model.predict_step, model.test_step)
+        if step_fn is not None
+    ):
+        rank_zero_warn(
+            "You are using the `dataloader_iter` step flavor. If you consume the iterator more than once per step, the"
```

Comment:
Shouldn't this warning suggest that you remove the batch_idx argument if you want to avoid the warning?

Reply:
No, it will still be a problem for all the hooks in the LM, LDM, or callbacks that also take the `batch_idx`. And since the user might not own their implementation it is a generally unavoidable issue